.github/workflows/build-wheels.yml

```diff
@@ -75,6 +77,12 @@ jobs:
         with:
           python-version: ${{ env.PYTHON_VERSION }}
 
+      - name: Set up QEMU
+        if: runner.os == 'Linux'
+        uses: docker/setup-qemu-action@v3
+        with:
+          platforms: all
+
```

Comment:
Thanks!
I suppose this doesn't run everything in QEMU, but only allows the container runtime to emulate a different CPU if needed, right?

Reply:
Thinking about it, I would prefer the same approach as for Apple Silicon builds here, for which we only cross-compile but skip the tests. I don't think a 5x CI time increase is really worth it.

Reply:
>Thanks!
I suppose this doesn't run everything in QEMU, but only allows the container runtime to emulate a different CPU if needed, right?


=> https://github.com/docker/setup-qemu-action , it seems to just install binaries that hook onto docker, making QEMU work transparently if a docker container shall be run which is not built for the current platform.

> prefer the same approach as for Apple Silicon builds here, for which we only cross-compile but skip the tests.

Let's take a look at the actual "Build" (collapsed wheel build + test) times that currently take place:

```
Build FastWARC

Building cp38-manylinux_aarch64 wheel
CPython 3.8 manylinux aarch64

Setting up build environment...
                                                              ✓ 0.88s
Building wheel...
                                                            ✓ 314.70s
Repairing wheel...
                                                              ✓ 7.99s
Testing wheel...
                                                             ✓ 66.51s
```

and

```
Build Resiliparse

Building cp38-manylinux_aarch64 wheel
CPython 3.8 manylinux aarch64

Setting up build environment...
                                                              ✓ 0.87s
Building wheel...
                                                            ✓ 424.65s
Repairing wheel...
                                                             ✓ 11.69s
Testing wheel...
                                                            ✓ 406.23s
```

Skipping Wheel Tests would only cut down on the CI times by ~35% or so, the large increase in general CI times would not be circumventable when also building for aarch64.

Reply:
I'm a bit unaware how to easily do that, if at all natively with manylinux. 

The official documentation states how to do cross-compilation without emulation of other architectures only on MacOS: https://cibuildwheel.readthedocs.io/en/stable/faq/#how-to-cross-compile (and even there it says they use Rosetta for building amd64 wheels on arm: "...for macOS Intel through Rosetta 2 emulation...")

How to build non-native Linux wheels is only described by doing emulation:  https://cibuildwheel.readthedocs.io/en/stable/faq/#emulation

Also other issues and posts on StackOverflow only mention QEMU: https://stackoverflow.com/questions/75806892/is-it-possible-to-build-an-arm64-linux-wheel-on-github

Reply:
We build Apple Silicon binaries on x86_64 and just skip the tests. This should also work with the x86_64 manylinux image as long as you install the aarch64 multiarch or multilib binaries of all dependencies.

Reply:
I agree that cross compilation is easily possible for Mac, as also stated in the documentation I linked above, because of (citation of the documentation):

> Cross-compilation is provided by the Xcode toolchain.

For cross-compilation on Linux using the official manylinux containers, it really seems to me that it's connected with more effort. I had a quick look at some packages I'm currently using and how they build wheels:

https://github.com/Ousret/charset_normalizer/blob/master/.github/workflows/cd.yml
https://github.com/contourpy/contourpy/blob/main/.github/workflows/build_wheels.yml
https://github.com/aio-libs/frozenlist/blob/master/.github/workflows/reusable-build-wheel.yml
https://github.com/pallets/markupsafe/blob/main/.github/workflows/publish.yaml
https://github.com/msgpack/msgpack-python/blob/main/.github/workflows/wheel.yml

They all exclusively do the same QEMU things that are also mentioned in the official manylinux documentation. I currently don't have the capacity to investigate how this would be possible without QEMU, I'm sorry. :/

Reply:
In theory, all it requires is installing the aarch64 versions of the dependency packages. I'll se if I find the time to test that out the next days. If not, I'll merge as is and we fix it later.

Reply:
The most prominent issue and PR to manylinux/cibuildwheel I could find for cross-compilation stems from authors of scikit-learn that had to do QEMU and we're unsatisfied with CI pipelines taking 6h:

- https://github.com/pypa/cibuildwheel/issues/598
- https://github.com/pypa/cibuildwheel/pull/804#issuecomment-907342905

Unfortunately, this is stale and/or closed because of various, deeply rooted issues.

What they seem to do currently is not using QEMU anymore but only doing native compilations, so *no* aarch64 linux wheel just from Github Actions as can be seen here:

- https://github.com/scikit-learn/scikit-learn/actions/runs/7295961757
- https://github.com/scikit-learn/scikit-learn/blob/main/.github/workflows/wheels.yml

The way that they still end up with aarch64 wheels in the end is that they use [CirrusCI](https://cirrus-ci.org/) to get their remaining, missing wheels:

- https://github.com/scikit-learn/scikit-learn/blob/main/build_tools/cirrus/arm_wheel.yml

This is a payed service offering native ARM Servers and Mac machines for direct integration into Github Actions.